src/VisualStudio/Core/Def/Implementation/TableDataSource/ProgressReporter.cs

```diff
@@ -0,0 +1,89 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.VisualStudio.LanguageServices.Implementation.Utilities;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.Shell.TableControl;
+using Microsoft.VisualStudio.Shell.TableManager;
+using Microsoft.VisualStudio.TaskStatusCenter;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.TableDataSource
+{
+    internal class ProgressReporter
+    {
+        private readonly IVsTaskStatusCenterService _taskCenterService;
+        private readonly TaskHandlerOptions _options;
+
+        private TaskCompletionSource<bool> _currentTask;
+        private ITaskHandler _taskHandler;
+        private string _lastMessage;
+
+        public ProgressReporter(string title, IVsTaskStatusCenterService taskCenterService)
+        {
+            _taskCenterService = taskCenterService;
+
+            _options = new TaskHandlerOptions()
+            {
+                Title = title,
+                ActionsAfterCompletion = CompletionActions.None
+            };
+        }
+
+        public void Started(bool started)
+        {
+            if (started)
+            {
+                // if there is any pending one. make sure it is finished.
+                _currentTask?.TrySetResult(true);
+
+                _taskHandler = _taskCenterService.PreRegister(_options, default);
+
+                _currentTask = new TaskCompletionSource<bool>();
+                _taskHandler.RegisterTask(_currentTask.Task);
```

Comment:
💭 I don't like that this introduces a new way to track progress, and then manually manages it. It would be better to link this either directly to a `Task` representing the overall asynchronous operation, or to an async token that already abstracts this.

Reply:
I am not sure what do you mean.

Reply:
I think I understand you now. unfortunately there is no original Task this can reuse. and AsyncToken is for different purpose. can't be used in this since AsyncToken is null object in production code.

only reason this need Task is because the underlying API this uses Task Status Center Service want to track work by Task. if they also supported traditional Start/Stop method, we won't need TaskCompleteionSource to use the API.

and TaskCompletionSource is explicitly there to convert traditional Start/Stop pattern to Task concept. and Task is the abstraction that present Start/Stop so not sure whether we need to create yet another abstract for exactly what TaskCompletionSource and Task are for.

Reply:
I could use TaskFactory.FromAsync to hide TaskCompletionSource but. then I need to deal with double call as well. so I like the way it is done now.